drivers/hmis/app/graphql/mutations/create_enrollment.rb

```diff
@@ -9,11 +9,11 @@ class CreateEnrollment < BaseMutation
     field :errors, [Types::HmisSchema::ValidationError], null: false
 
     def validate_input(project_id:, start_date:, household_members:)
-      errors = []
-      errors << InputValidationError.new('Exactly one client must be head of household', attribute: 'relationship_to_ho_h') if household_members.select { |hm| hm.relationship_to_ho_h == 1 }.size != 1
-      errors << InputValidationError.new('Entry date cannot be in the future', attribute: 'start_date') if Date.parse(start_date) > Date.today
-      errors << InputValidationError.new("Project with id '#{project_id}' does not exist", attribute: 'project_id') unless Hmis::Hud::Project.editable_by(current_user).exists?(id: project_id)
-      errors
+      errors = HmisErrors::CustomValidationErrors.new
+      errors.add :relationship_to_ho_h, full_message: 'Exactly one client must be head of household' if household_members.select { |hm| hm.relationship_to_ho_h == 1 }.size != 1
+      errors.add :start_date, :out_of_range, message: 'cannot be in the future', readable_attribute: 'Entry date' if Date.parse(start_date) > Date.today
```

Comment:
I wonder if we should add a warning if the start date is more than a year (or maybe 9 months) in the past.  People often accidentally add an entry date on the DOB, and I bet in January, there are a bunch of `1/5/2022` that should be `1/5/2023`.  Also, does it make sense to collect up the validations for certain fields (entry date, exit date, information date, come to mind) in one place so we can say, "this is an entry date, and an error object, figure out what might be wrong with it and add the errors and warnings."  Something like:
```
errors, warnings = HmisErrors::CustomValidationErrors.validate_field(:entry_date, input_name: :start_date, value: start_date, errors: errors, warnings: warnings)
.....
# in HmisErrors::CustomValidationErrors
def  self.validate_field(field_name, input_name:, value:, errors:, warnings:)
   case field_name
     when :entry_date
       errors.add(input_name, out_of_range, message: 'cannot be in the future', readable_attribute: 'Entry date') if Date.parse(value) > Date.today
      warnings.add(input_name, really_old, message: 'are you sure you meant to add an entry date more than 9 months old', readable_attribute: 'Entry date') if Date.parse(value) < 9.months.ago.to_date
      warnings.add(input_name, matches_dob, message: 'did you intend to enroll the client on their date of birth?  This is really uncommon', readable_attribute: 'Entry date') if Date.parse(value) == <magic to obtain dob>
   end
   [errors, warnings]
end
```

I'm not quite happy with the setup, but I haven't fully dug into the API yet, so maybe there's a cleaner way, like just setting the warning flag on the error object instead of passing in both errors and warnings and then restructuring the return, but I think this will help keep the validation logic in one place for the bits that show up on multiple forms.

Reply:
Looks like `severity: :warning` should be all you need for the warnings, so that could be dropped from the arguments and return.

Reply:
Thinking about this a bit more, these could even delegate the error messages down to the class that should be responsible for them.  So you might want something more like
```
# in HmisErrors::CustomValidationErrors
def  self.validate_field(field_name, input_name:, value:, errors:)

   case field_name
     when :entry_date
       Hmis::Hud::Enrollment.validate_entry_date(value).each do |error_type, error|
         errors.add(input_name, error_type, **error)  
       end
   end
   errors
end
```
and 
```
# in Hmis::Hud::Enrollment
def self.validate_entry_date(value)
  errors = []
  errors <<  { out_of_range: { message: 'cannot be in the future', readable_attribute: 'Entry date' }} if Date.parse(value) > Date.today
   errors <<  { really_old: { message: 'are you sure you meant to add an entry date more than 9 months old', readable_attribute: 'Entry date', severity: 'warning' }}  if Date.parse(value) < 9.months.ago.to_date
   errors <<  { matches_dob:  { message: 'did you intend to enroll the client on their date of birth?  This is really uncommon', readable_attribute: 'Entry date' }} if Date.parse(value) == <magic to obtain dob>
  errors
end

Reply:
Not sure if this would even be possible, but potentially, you could just call validate on a new object and retrieve the relevant errors from the ActiveRecord errors.  Something like:
```
en = Hmis::Hud::Enrollment.new(entry_date: value)
en.valid?
en.errors.each do |error|
  next unless error.attribute == :entry_date

  errors.add(...)
end
```

Reply:
Created https://www.pivotaltracker.com/story/show/184407413